cmd/kuberhealthy/kuberhealthy.go

```diff
@@ -1292,20 +1291,46 @@ func (k *Kuberhealthy) externalCheckReportHandler(w http.ResponseWriter, r *http
 	// make a request ID for tracking this request
 	requestID := "web: " + uuid.New().String()
 
-	k.externalCheckReportHandlerLog(requestID, "Client connected to check report handler from", r.RemoteAddr, r.UserAgent())
+	k.externalCheckReportHandlerLog(requestID, "Client connected to check report handler from", r.UserAgent())
 
-	// validate the calling pod to ensure that it has a proper KH_CHECK_NAME and KH_RUN_UUID
-	k.externalCheckReportHandlerLog(requestID, "validating external check status report from: ", r.RemoteAddr)
-	ipReport, err := k.validateExternalRequest(r.RemoteAddr)
-	if err != nil {
-		w.WriteHeader(http.StatusBadRequest)
-		k.externalCheckReportHandlerLog(requestID, "Failed to look up pod by IP:", r.RemoteAddr, err)
-		return nil
+	// Get kh-run-uuid from the request header and verify it exists. If the kh-run-uuid doesn't exist, verify request by
+	// validating using the pod's remote IP.
+	reportingKHRunUUID := r.Header.Get("kh-run-uuid")
+	var podReport PodReportInfo
+	var err error
+	if len(reportingKHRunUUID) == 0 {
```

Comment:
function all this logic out into something like this:

```go
var reportValidated bool
var podReport PodReportInfo
var err error
if requestUsesUUIDHeader(r) {
      podReport, err = ValidateCheckUUID(r.Header.Get("kh-run-uuid"))
      if err != nil {
          // log the error we had validating
      } else {
          // if no issue validating, set a bool so that we don't try the legacy IP resolution checks
          checkValidated = true
      }
}
```

Then, drop the else below this code for something like this:

```go
if !reportValidated {
  // try to validate pod by searching for its source it (legacy behavior)
  podReport, err = ValidateCheckUUIDBySourceIP(r)
  if err != nil {
      // throw error back to client and log it. both header and source ip validation have failed
  }
}
```

Reply:
Just remember that we want to try to validate with the header first because it does not require the extra call to the Kube API.  We fall back to doing a query for pods by IP _only_ if the header validation fails.